### PR TITLE
test(flat-table): typescript refactor

### PR DESCRIPTION
**Pull request by stephenogorman**
### Proposed behaviour

- Rename the `flat-table.cy.js` to `flat-table.cy.tsx` file in `./cypress/components/flat-table/` folder.
- Refactor tests to Typescript.  

### Current behaviour

`Flat-table` tests are in js not ts.

### Checklist

- [x] Commits follow our style guide
- [ ] Related issues linked in commit messages if required
- [ ] Screenshots are included in the PR if useful
- [ ] All themes are supported if required
- [ ] Unit tests added or updated if required
- [x] Cypress automation tests added or updated if required
- [ ] Storybook added or updated if required
- [ ] Translations added or updated (including creating or amending translation keys table in storybook) if required
- [ ] Typescript `d.ts` file added or updated if required
- [ ] Related docs have been updated if required

#### QA

- [x] Tested in CodeSandbox/storybook
- [x] Add new Cypress test coverage if required
- [x] Carbon implementation matches Design System/designs
- [ ] UI Tests GitHub check reviewed if required

### Testing instructions
- [x] Run `npx cypress open --component` to check if the `dialog-full-screen.cy.tsx` file passed
- [x] Run `npx cypress run --component` to check none of the other `*.cy.tsx` files have regressed

### CODE REVIEW

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 1]
```diff
@@ -0,0 +1,2681 @@
+/* eslint-disable jest/no-disabled-tests */
+import React from "react";
+import {
+  FlatTableBodyDraggableProps,
+  FlatTableCheckboxProps,
+} from "components/flat-table";
```
Comment (DlgSHi) [inline cypress/components/flat-table/flat-table.cy.tsx:6]: could you update the path to be like below?

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 2]
```diff
@@ -0,0 +1,2681 @@
...
+        );
+
+        flatTableWrapper().should("have.css", "width").as("wrapperWidth");
+        cy.get("@wrapperWidth")
+          .then(($el) => parseInt($el.toString()))
+          .should("be.within", width - 1, width + 1);
+      }
+    );
+
+    it.each([["visible"], ["hidden"], ["clip"], ["scroll"], ["auto"]])(
```
Comment (DlgSHi) [inline cypress/components/flat-table/flat-table.cy.tsx:448]:
```suggestion
    it.each(["visible", "hidden", "clip", "scroll", "auto"])(
```

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 3]
```diff
@@ -0,0 +1,2681 @@
...
+        .first()
+        .should("have.css", "border-radius", "8px 0px 0px");
+      flatTableHeaderCells().last().should("have.css", "border-radius", "0px");
+      flatTableCell(16).should("have.css", "border-radius", "0px");
+      flatTableCell(19).should("have.css", "border-radius", "0px");
+      flatTablePager().should("have.css", "border-radius", "0px");
+    });
+
+    it("has the expected border radius styling when horizontal scrollbar exists", () => {
+      CypressMountWithProviders(
```
Comment (DlgSHi) [inline cypress/components/flat-table/flat-table.cy.tsx:2582]: why we can't move this component to `flat-table-test.stories.tsx` file?

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 4]
```diff
@@ -0,0 +1,2681 @@
...
+          </FlatTableBody>
+        </FlatTable>
+      );
+
+      flatTableCell(6).should("have.css", "border-radius", "0px");
+      flatTableCell(11).should("have.css", "border-radius", "0px");
+    });
+
+    it("has the expected border radius styling when first column has rowspan", () => {
+      CypressMountWithProviders(
```
Comment (DlgSHi) [inline cypress/components/flat-table/flat-table.cy.tsx:2624]: same question here

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 5]
```diff
@@ -0,0 +1,2681 @@
...
+        </FlatTable>
+      );
+
+      flatTableCell(0).should("have.css", "border-radius", "0px 0px 0px 8px");
+      flatTableCell(3).should("have.css", "border-radius", "0px");
+      flatTableCell(4).should("have.css", "border-radius", "0px 0px 8px");
+    });
+
+    it("has the expected border radius styling when last column has rowspan", () => {
+      CypressMountWithProviders(
```
Comment (DlgSHi) [inline cypress/components/flat-table/flat-table.cy.tsx:2653]: and here

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 6]
```diff
@@ -397,3 +459,2716 @@ export const ExpandableWithLink = () => {
     </FlatTable>
   );
 };
+
+export const FlatTableComponent = ({ ...props }) => {
```
Comment (DlgSHi) [inline src/components/flat-table/flat-table-test.stories.tsx:463]:
```suggestion
export const FlatTableComponent = (props: FlatTableProps) => {
```

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 7]
```diff
@@ -397,3 +459,2716 @@ export const ExpandableWithLink = () => {
...
+            <FlatTableCell>Single</FlatTableCell>
+            <FlatTableCell>0</FlatTableCell>
+          </FlatTableRow>
+        </FlatTableBody>
+      </FlatTable>
+    </div>
+  );
+};
+
+export const FlatTableFooterComponent = ({ ...props }) => {
```
Comment (DlgSHi) [inline src/components/flat-table/flat-table-test.stories.tsx:554]: typing for `props`

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 8]
```diff
@@ -397,3 +459,2716 @@ export const ExpandableWithLink = () => {
...
+            <FlatTableCell>5</FlatTableCell>
+          </FlatTableRow>
+        </FlatTableBody>
+      </FlatTable>
+    </div>
+  );
+};
+
+export const FlatTableColorRowSelectableComponent = (
+  props: Partial<FlatTableCheckboxProps> & Partial<FlatTableCheckboxProps>
```
Comment (DlgSHi) [inline src/components/flat-table/flat-table-test.stories.tsx:1265]: same props passed

```suggestion
  props: Partial<FlatTableCheckboxProps>
```

**Review by DlgSHi — CHANGES_REQUESTED** [review]

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 9]
```diff
@@ -0,0 +1,2681 @@
+/* eslint-disable jest/no-disabled-tests */
```
Comment (divyajindel) [inline cypress/components/flat-table/flat-table.cy.tsx:1]: Could you please remove this line ?

Comment (divyajindel) [inline cypress/components/flat-table/flat-table.cy.tsx:1]: Could you please remove this line ?

**Review by divyajindel — CHANGES_REQUESTED** [review]

File: src/components/flat-table/flat-table-test.stories.tsx [hunk 10]
```diff
@@ -397,3 +459,2716 @@ export const ExpandableWithLink = () => {
...
+            </FlatTableCell>
+          </FlatTableRow>
+        </FlatTableBody>
+      </FlatTable>
+    </div>
+  );
+};
+
+export const FlatTableSortingComponent = (
+  props: Partial<FlatTableProps> & Partial<FlatTableCellProps>
```
Comment (divyajindel) [inline src/components/flat-table/flat-table-test.stories.tsx:1841]:
```suggestion
 props: Partial<FlatTableProps> & FlatTableCellProps
```

**Review by divyajindel — CHANGES_REQUESTED** [review]

**Review by divyajindel — APPROVED** [review]

File: cypress/components/flat-table/flat-table.cy.tsx [hunk 11]
```diff
@@ -0,0 +1,2595 @@
...
+      }
+    );
+
+    it("should render Flat Table with rowSpan set to make header cells span 2 rows", () => {
+      CypressMountWithProviders(
+        <stories.FlatTableSpanComponent width="500px" />
+      );
+
+      for (let i = 0; i < 4; i++) {
+        if (i === 0 || i === 1 || i === 3) {
```
Comment (DipperTheDan) [inline cypress/components/flat-table/flat-table.cy.tsx:456]: **suggestion:** could do the inverse here and say: 
```suggestion
        if (i !== 2 || i !== 4) {
```
uses less parameters in the `if` condition. I'm happy either way though and it's fine this way if this is how you lot in QA want the tests to be written.

**Review by DipperTheDan — APPROVED** [review]

**Review by DlgSHi — APPROVED** [review]

**Review by divyajindel — APPROVED** [review]

**Pull request merged**